### ISSUE TITLE
RegenerateMapException Should Mention to Use ConfigAndGenerateSafe

### ISSUE THREAD

**Issue by DavidFidge** [Chris3606/GoRogue#253]
develop-3.0 branch as at [8af4253d884e72490467fb1ecf5a0d56b53545c4](https://github.com/Chris3606/GoRogue/commit/8af4253d884e72490467fb1ecf5a0d56b53545c4)

Reproducible by creating a Generator, adding the DungeonMazeMapSteps step algorithm and performing Generate (see test case below).  Only happens about once every 500 executions.  

```
        [TestMethod]
        public void Generate_Should_Work()
        {
            for (var i = 0; i < 100000; i++)
            {
                var generator = new Generator(50, 50);

                generator.AddSteps(DefaultAlgorithms.DungeonMazeMapSteps());

                generator.Generate();
            }
        }
```

GoRogue.MapGeneration.RegenerateMapException
  HResult=0x80131500
  Message=RoomDoorConnection was given rooms that already had doors.
  Source=GoRogue
  StackTrace:
   at GoRogue.MapGeneration.Steps.RoomDoorConnection.<OnPerform>d__10.MoveNext() in C:\dev\GoRogue\GoRogue\MapGeneration\Steps\RoomDoorConnection.cs:line 203
   at GoRogue.MapGeneration.GenerationStep.PerformStep(GenerationContext context) in C:\dev\GoRogue\GoRogue\MapGeneration\GenerationStep.cs:line 237
   at GoRogue.MapGeneration.Generator.Generate() in C:\dev\GoRogue\GoRogue\MapGeneration\Generator.cs:line 132

**Comment by Chris3606** [comment]
Hi.  Thank you for the bug report!  As the API documentation for the exception indicates, however, this is known (and sometimes expected) behavior.  Some map generators cannot avoid invalid states without a large performance hit or while still maintaining some semblance of bounded runtime, so instead they raise an exception indicating that they need to start over.

This exception is designed to be thrown and automatically handled via the `Generator.ConfigAndGenerateSafe` function.  If you change your source code to this, it should produce identical results but without that error:

```CS
[TestMethod]
        public void Generate_Should_Work()
        {
            for (var i = 0; i < 100000; i++)
            {
                var generator = new Generator(50, 50);

                generator.ConfigAndGenerateSafe(gen => gen.AddSteps(DefaultAlgorithms.DungeonMazeMapSteps());
            }
        }
```

For reference, `ConfigAndGenerateSafe` function effectively just detects when this exception occurs, and restarts the map generation process from scratch; so if you prefer to do this manually in your code, this should also work.

Ideally, I'd like to remove the need for this exception from the framework entirely, but because it's always possible (though sometimes very unlikely) to run into cases where RNG leads to unfavorable conditions, I haven't been able to justify removing the exception.  This particular instance of the exception is actually covering up a limitation in the algorithm used to ensure that there is a single maze which is always connected to itself between the rooms.  The connectivity algorithm can't guarantee that it won't accidentally cut through a room trying to connect different mazes (although it won't do this very often); and if it does this, it breaks assumptions used to enforce the parameters of door placement; so this works around that.

Bottom line; please feel free to let me know if the fixed code doesn't work for you; and if it does, where could I have documented this so that it would have saved you the trouble?

**Comment by Chris3606** [comment]
Per the discussion on Discord, I'll change the exception message for `RegenerateMapException` to indicate that you should consider using `ConfigAndGenerateSafe` to reinforce the API documentation.
